### PR TITLE
document missing keywords

### LINKED ISSUE #1403: Solidity keywords based on grammar.txt
The following keywords should not be used outside their context defined in grammar.txt, for example not as identifiers. Else-wise Solidity code can not be processed by lexical analysers based on lex-like programs.  
 
```
-define(TOKEN_PATTERNS, [
    {"^(?i)(ADDRESS)$",          'ADDRESS'},
    {"^(?i)(ANONYMOUS)$",        'ANONYMOUS'},
    {"^(?i)(AS)$",               'AS'},
    {"^(?i)(BOOL)$",             'BOOL'},
    {"^(?i)(BREAK)$",            'BREAK'},
    {"^(?i)(CONSTANT)$",         'CONSTANT'},
    {"^(?i)(CONTINUE)$",         'CONTINUE'},
    {"^(?i)(CONTRACT)$",         'CONTRACT'},
    {"^(?i)(DAYS)$",             'DAYS'},
    {"^(?i)(DELETE)$",           'DELETE'},
    {"^(?i)(DO)$",               'DO'},
    {"^(?i)(ELSE)$",             'ELSE'},
    {"^(?i)(ENUM)$",             'ENUM'},
    {"^(?i)(ETHER)$",            'ETHER'},
    {"^(?i)(EVENT)$",            'EVENT'},
    {"^(?i)(EXTERNAL)$",         'EXTERNAL'},
    {"^(?i)(FALSE)$",            'FALSE'},
    {"^(?i)(FINNEY)$",           'FINNEY'},
    {"^(?i)(FOR)$",              'FOR'},
    {"^(?i)(FROM)$",             'FROM'},
    {"^(?i)(FUNCTION)$",         'FUNCTION'},
    {"^(?i)(HOURS)$",            'HOURS'},
    {"^(?i)(IF)$",               'IF'},
    {"^(?i)(IMPORT)$",           'IMPORT'},
    {"^(?i)(INDEXED)$",          'INDEXED'},
    {"^(?i)(INTERNAL)$",         'INTERNAL'},
    {"^(?i)(IS)$",               'IS'},
    {"^(?i)(LIBRARY)$",          'LIBRARY'},
    {"^(?i)(MAPPING)$",          'MAPPING'},
    {"^(?i)(MEMORY)$",           'MEMORY'},
    {"^(?i)(MINUTES)$",          'MINUTES'},
    {"^(?i)(MODIFIER)$",         'MODIFIER'},
    {"^(?i)(NEW)$",              'NEW'},
    {"^(?i)(PRAGMA)$",           'PRAGMA'},
    {"^(?i)(PRIVATE)$",          'PRIVATE'},
    {"^(?i)(PUBLIC)$",           'PUBLIC'},
    {"^(?i)(RETURN)$",           'RETURN'},
    {"^(?i)(RETURNS)$",          'RETURNS'},
    {"^(?i)(SECONDS)$",          'SECONDS'},
    {"^(?i)(STORAGE)$",          'STORAGE'},
    {"^(?i)(STRING)$",           'STRING'},
    {"^(?i)(STRUCT)$",           'STRUCT'},
    {"^(?i)(SZABO)$",            'SZABO'},
    {"^(?i)(THROW)$",            'THROW'},
    {"^(?i)(TRUE)$",             'TRUE'},
    {"^(?i)(USING)$",            'USING'},
    {"^(?i)(VAR)$",              'VAR'},
    {"^(?i)(WEEKS)$",            'WEEKS'},
    {"^(?i)(WEI)$",              'WEI'},
    {"^(?i)(WHILE)$",            'WHILE'},
    {"^(?i)(YEARS)$",            'YEARS'}
]).
```

### PR DESCRIPTION
**Pull request by pirapira**
I added `null` and `default` as reserved keywords.  I added `assembly` in `grammar.txt` together with the inline assembly grammar.

This fixes #1403.

### CODE REVIEW

File: docs/grammar.txt [hunk 1]
```diff
@@ -54,13 +54,52 @@ WhileStatement = 'while' '(' Expression ')' Statement
...
+
+OneArgInstruction = 'iszero' | 'not' | 'balance' | 'calldatalload' | 'extcodesize' | 'blockhash' | 'pop' | 'mload' | 'sload' | 'jump' | 'suicide'
+
+TwoArgInstruction = 'add' | 'mul' | 'sub' | 'div' | 'sdiv' | 'mod' | 'smod' | 'exp' | 'signextend' | 'lt' | 'gt' | 'slt' | 'sgt' | 'eq' | 'and' | 'or' | 'xor' | 'byte' | 'sha3' | 'jumpi' | 'mstore' | 'mstore8' | 'sstore' | 'log0' | 'return'
+
+ThreeArgInstruction = 'addmod' | 'mulmod' | 'calldatacopy' | 'codecopy' | 'log1'
+
+FourArgInstruction = 'extcodecopy' | 'log2'
+
+FiveArginstruction = 'log3'
```
Comment (axic) [inline docs/grammar.txt]: I think this should be with a capital `Instruction`.

File: docs/grammar.txt [hunk 2]
```diff
@@ -54,13 +54,52 @@ WhileStatement = 'while' '(' Expression ')' Statement
...
+
+SixArgInstruction = 'log4' | 'delegatecall'
+
+SevenArginstruction = 'call' | 'callcode'
```
Comment (axic) [inline docs/grammar.txt]: Same as above.

**Comment by axic** [comment]
I think we should only declare the syntax for assembly instructions, i.e. keywords and parenthesis.

Having the list there opens up a bigger problem when we have macros available in inline assembly - do we need to define every single one of them in the grammar?

File: docs/grammar.txt [hunk 3]
```diff
@@ -54,13 +54,52 @@ WhileStatement = 'while' '(' Expression ')' Statement
...
 DoWhileStatement = 'do' Statement 'while' '(' Expression ')'
 Continue = 'continue'
 Break = 'break'
 Return = 'return' Expression?
 Throw = 'throw'
 VariableDefinition = VariableDeclaration ( '=' Expression )?
 
+InlineAssemblyBlock = '{' AssemblyItem* '}'
+
+AssemblyItem = SingleAssemblyInstruction | FunctionalAssemblyExpression | InlineAssemblyBlock | AssemblyLocalBinding | AssemblyAssignment | Identifier | NumbeerLiteral | StringLiteral | HexLiteral
```
Comment (chriseth) [inline docs/grammar.txt]: Numbeer

File: docs/grammar.txt [hunk 4]
```diff
@@ -54,13 +54,52 @@ WhileStatement = 'while' '(' Expression ')' Statement
...
+| OneArgInstruction '(' FunctionalAssemblyExpression ')'
+| TwoArgInstruction '(' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ')'
+| ThreeArgInstruction '(' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ')'
+| FourArgInstruction '(' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ')'
+| FiveArgInstruction '(' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ')'
+| SixArgInstruction '(' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ')'
+| SevenArgInstruction '(' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ',' FunctionalAssemblyExpression ')'
+
+AssemblyAssignment = Identifier ':=' FunctionalAssemblyExpression | '=:' Identifier
```
Comment (chriseth) [inline docs/grammar.txt]: Probably better to move this up. In general, I think the `XArgInstruction` should go to the end.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: docs/grammar.txt (hunk 3 above)
Comment (pirapira) [reply docs/grammar.txt]: Fixed.

File: docs/grammar.txt (hunk 4 above)
Comment (pirapira) [reply docs/grammar.txt]: I moved the `AssemblyAssignment` item above; also I moved the whole inline assembly syntax at the end.

File: docs/grammar.txt (hunk 1 above)
Comment (pirapira) [reply docs/grammar.txt]: Fixed.

File: docs/grammar.txt (hunk 2 above)
Comment (pirapira) [reply docs/grammar.txt]: Fixed.

**Comment by pirapira** [comment]
@axic I have a secret plan of following these definitions to generate inline assembly code randomly.  For macros I would just say `Identifier` followed by some arguments.

**Comment by axic** [comment]
> For macros I would just say `Identifier` followed by some arguments.

The instructions aren't different from that though.

**Comment by pirapira** [comment]
@axic That's a fair point.  I searched for `send` `call` and `deletetecall` but these don't get special treatments in `grammar.txt` either.  I'll just say `Identifier` followed by parameter list.

**Comment by pirapira** [comment]
@axic done.

**Comment by axic** [comment]
It must be noted though there three exceptions in our parser: https://github.com/ethereum/solidity/blob/develop/libsolidity/inlineasm/AsmParser.cpp#L156

`return`, `byte`, `address` is parsed as a keyword and not as an identifier as it needs to be amended for inline assembly.

File: docs/grammar.txt (hunk 4 above)
Comment (pirapira) [reply docs/grammar.txt]: Even better, now `XArgInstruction`s disappeared.

**Comment by pirapira** [comment]
@axic `grammar.txt` does not exclude keywords from the possible identifier.

**Comment by walter-weinmann** [comment]
I think it is quite normal that a grammar specification does not explicitly specify that the use of keywords as an identifier is not allowed.

**Review by chriseth — APPROVED** [review]

**Pull request merged**